chainer/functions/array/where.py

```diff
@@ -16,30 +17,33 @@ def check_type_forward(self, in_types):
         type_check.expect(
             c_type.dtype == numpy.bool_,
             x_type.dtype == y_type.dtype,
-            x_type.shape == c_type.shape,
-            y_type.shape == c_type.shape,
         )
+        type_check.expect_broadcast_shapes(
+            c_type.shape, x_type.shape, y_type.shape)
 
     def forward(self, inputs):
+        # may broadcast
         self.retain_inputs((0,))
         xp = cuda.get_array_module(*inputs)
         condition, x, y = inputs
+        self.x_shape = x.shape
+        self.y_shape = y.shape
         return xp.where(condition, x, y),
 
     def backward(self, indexes, grad_outputs):
         condition = self.get_retained_inputs()[0]
         xp = cuda.get_array_module(condition.data)
         g, = grad_outputs
-        zeros = xp.zeros(g.shape, dtype=g.dtype)
+        zero = xp.zeros((), dtype=g.dtype)
         ret = []
         if 0 in indexes:
             ret.append(None)
```

Comment:
It's not related to this PR, but I found that the first argument (condition) could be an `__init__` parameter instead of the `forward` argument, then there is no need of constructing Variable object for it (and no need to retain). It's ok to leave it in this PR.